src/mpid/ch4/netmod/ofi/ofi_types.h

```diff
@@ -397,6 +350,23 @@ typedef struct {
     char kvsname[MPIDI_KVSAPPSTRLEN];
     char pname[MPI_MAX_PROCESSOR_NAME];
     int port_name_tag_mask[MPIR_MAX_CONTEXT_MASK];
+
+    /* Capability settings */
+#ifdef MPIDI_OFI_ENABLE_RUNTIME_CHECKS
```

Comment:
I'm not sure we need something that complicated.  Why not simply search through the available capability sets with those hardcoded by the user, and use the first one that matches.

Minor comment: can you change `runtime` to `any`, since it's really asking MPICH to pick any provider that it can find.

With respect to testing for the capability sets, I think you are considering two types of CVARs:

1. The first type is a single CVAR to pick the provider through a CVAR, say `MPIR_CVAR_OFI_PROVIDER=psm2`.  I think this is easy to enable.

2. The second type is a collection of CVARs that allow the user to pick each capability specifically (e.g., I want to enable/disable tagged communication using `MPIR_CVAR_OFI_TAGGED=0`).  I think this is possible too.

Here are some possible cases:

1. Let's say both `MPIR_CVAR_OFI_PROVIDER` and `MPIR_CVAR_OFI_TAGGED` are on their default values (which are "NULL" and "-1").  When you do `fi_info`, let's say the provider you got 3 capability sets from sockets (s1, s2, s3), and two from psm2 (p1, p2).  Now suppose sockets did not specify any hardcoded capability set.  Then, you simply use the first socket capability set (s1).

2. Same case as above, but the user set `MPIR_CVAR_OFI_PROVIDER=psm2`, then you pass that provider in your `fi_info` call to only get psm2 specific capability sets.  Now, for psm2, let's say you have hardcoded capability sets available: here you look through the available capability sets and find the ones that give you the hardcoded capability set.

3. Same as (2) above, but let's say that `MPIR_CVAR_OFI_TAGGED=0` was set.  Then you simply comb through your capability sets and try to find one with the above value for `tagged`.